File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel

    # Consumer(WebSocketのクライアント)がChannelをsubscribe（受信）することでsubscriberとして振る舞う。

  def subscribed
    stream_from "room_channel_#{params['room']}"
    # Streamすることで、Channelにルーティング機能を与える。
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    # jsで実行されたspeakのmessageを受け取り、room_channelのreceivedにブロードキャストする
    # ActionCable.server.broadcast 'room_channel', message: data['message']
    Message.create! content: data['message'], user_id: current_user.id, room_id: params['room']
  end
end

File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

$(function() {
  const chatChannel = consumer.subscriptions.create({channel: "RoomChannel", room: $('#message').data('room_id') }, {
    connected() {
      // Called when the subscription is ready for use on the server
    },
  
    disconnected() {
      // Called when the subscription has been terminated by the server
    },
  
    // データ受信。room_channel.rbでブロードキャストされたものが届く
    received: function(data) {
      return $('#message').append(data['message']);
      // Called when there's incoming data on the websocket for this channel
    },
    // これが実行されるとコンシューマ(WebSocket Connectionのクライアント))になったRoomChannel#speak({ message: message })が呼ばれる
    speak: function(message) {
      return this.perform('speak', {
        message: message
      });
    }
  });
  
  $(document).on('keypress', '[data-behavior~=room_speaker]', function(event) {
    //keycodeが13(すなわちEnterキー)を押された時の条件分岐
    if (event.keyCode === 13) {
      chatChannel.speak(event.target.value);
      event.target.value = '';
      return event.preventDefault();
    }
  });
});




